geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java

```diff
@@ -120,6 +123,60 @@
 
   private static int bridgeServerPort;
 
+  @Test
+  public void testPutAllWithCQLocalDestroy() throws Exception {
+    VM server1 = getVM(0);
+    VM server2 = getVM(1);
+    VM client = getVM(2);
+
+    final String cqName = "testPutAllWithCQLocalDestroy_0";
+    createServer(server1);
+    createServer(server2);
+    final String host = Host.getHost(0).getHostName();
+    final int port = server2.invoke(() -> PartitionedRegionCqQueryDUnitTest.getCacheServerPort());
+    createClient(client, port, host);
+    createCQ(client, cqName, cqs[0]);
+
+    int numObjects = 1000;
+
+    server1.invoke(() -> {
+      Region region = getCache().getRegion(SEPARATOR + "root" + SEPARATOR + regions[0]);
+      Map buffer = new HashMap();
+      for (int i = 1; i < numObjects; i++) {
+        Portfolio p = new Portfolio(i);
+        buffer.put("" + i, p);
+      }
+      region.putAll(buffer);
+    });
+
+    client.invoke(() -> {
+      QueryService cqService = getCache().getQueryService();
+      CqQuery cqQuery = cqService.getCq(cqName);
+      if (cqQuery == null) {
+        fail("Failed to get CQ " + cqName);
+      }
+      try {
+        cqQuery.executeWithInitialResults();
+      } catch (Exception ex) {
+        fail("Failed to execute  CQ " + cqName, ex);
+      }
+    });
+
+    server1.invoke(() -> {
+      Region region = getCache().getRegion(SEPARATOR + "root" + SEPARATOR + regions[0]);
+      Map buffer = new HashMap();
+      for (int i = 1; i < numObjects; i++) {
+        Portfolio p = new Portfolio(-1 * i);
+        buffer.put("" + i, p);
+      }
+      region.putAll(buffer);
+    });
+
+    cqHelper.closeClient(client);
+    cqHelper.closeServer(server2);
+    cqHelper.closeServer(server1);
```

Comment:
It's better (and more industry standard for writing JUnit tests) if you can figure out a way to move cleanup code at the end of the test method to a `tearDown` method annotated with `@After`. For example, if you make `cqHelper.closeClient` return without error if client is null or not running, then you could hoist these local variables to fields and the test would look more like this:
```
private VM client;
private VM server1;
private VM server2;

@Before 
public void setUp() {
 server1 = getVM(0);
 server2 = getVM(1);
 client = getVM(2);
}

@After
public void tearDown() {
  cqHelper.closeClient(client);
  cqHelper.closeServer(server2);
  cqHelper.closeServer(server1);
}
```
And `cqHelper` close methods could have some early-outs like this:
```
public void closeClient(VM vm) {
  if (vm == null) {
    return;
  }
  // etc
}
```

Reply:
I would love the clean the test code as you suggested. This test also uses a lot of already deprecated test classes and APIs. It's better to refactor it. However, introducing @Before and @After would affect all 14 test methods in the file, it would be better to have a separate JIRA to track the refactoring. I have created https://issues.apache.org/jira/browse/GEODE-9020 to track it.

Reply:
Feel free to fit into the test in the best way you can. Another alternative is to extract the new test out to a new test class (now in a followup PR). You could also modernize the whole test but that would be best done as a followup PR.